rcl/test/rcl/test_expand_topic_name.cpp

```diff
@@ -0,0 +1,221 @@
+  rcutils_string_map_t subs = rcutils_get_zero_initialized_string_map();
+  rcutils_ret_t rcu_ret = rcutils_string_map_init(&subs, 0, allocator);
+  ASSERT_EQ(RCUTILS_RET_OK, rcu_ret);
+  ret = rcl_get_default_topic_name_substitutions(&subs);
+  ASSERT_EQ(RCL_RET_OK, ret);
+
+  const char * topic = "{node}/chatter";
+  const char * ns = "/my_ns";
+  const char * node = "my_node";
+  char * expanded_topic;
+
+  // pass null for topic string
+  {
+    ret = rcl_expand_topic_name(NULL, node, ns, &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+  }
+
+  // pass null for node name
+  {
+    ret = rcl_expand_topic_name(topic, NULL, ns, &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+  }
+
+  // pass null for node namespace
+  {
+    ret = rcl_expand_topic_name(topic, node, NULL, &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+  }
+
+  // pass null for subs
+  {
+    ret = rcl_expand_topic_name(topic, node, ns, NULL, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+  }
+
+  // pass null for expanded_topic
+  {
+    ret = rcl_expand_topic_name(topic, node, ns, &subs, allocator, NULL);
+    EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+    rcl_reset_error();
+  }
+
+  // pass invalid topic
+  {
+    ret = rcl_expand_topic_name("white space", node, ns, &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_TOPIC_NAME_INVALID, ret);
+    rcl_reset_error();
+  }
+
+  // pass invalid node name
+  {
+    ret = rcl_expand_topic_name(topic, "/invalid_node", ns, &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_NODE_INVALID_NAME, ret);
+    rcl_reset_error();
+  }
+
+  // pass invalid node namespace
+  {
+    ret = rcl_expand_topic_name(topic, node, "white space", &subs, allocator, &expanded_topic);
+    EXPECT_EQ(RCL_RET_NODE_INVALID_NAMESPACE, ret);
+    rcl_reset_error();
+  }
+}
+
+TEST(test_expand_topic_name, various_valid_topics) {
+  rcl_ret_t ret;
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcutils_string_map_t subs = rcutils_get_zero_initialized_string_map();
+  rcutils_ret_t rcu_ret = rcutils_string_map_init(&subs, 0, allocator);
+  ASSERT_EQ(RCUTILS_RET_OK, rcu_ret);
+  ret = rcl_get_default_topic_name_substitutions(&subs);
+  ASSERT_EQ(RCL_RET_OK, ret);
+
+  std::vector<std::vector<std::string>> topics_that_should_expand_to = {
+    // {"input_topic", "node_name", "/namespace", "expected result"},
+    {"/chatter", "my_node", "/my_ns", "/chatter"},
+    {"chatter", "my_node", "/my_ns", "/my_ns/chatter"},
+    {"{node}/chatter", "my_node", "/my_ns", "/my_ns/my_node/chatter"},
+    {"/{node}", "my_node", "/my_ns", "/my_node"},
+    {"{node}", "my_node", "/my_ns", "/my_ns/my_node"},
+    {"{ns}", "my_node", "/my_ns", "/my_ns"},
+    {"{namespace}", "my_node", "/my_ns", "/my_ns"},
```

Comment:
can you add a test with double substitutions?
`"{namespace}/{node}/chatter", "my_node", "/my_ns", "/my_ns/my_node/chatter"}`